src/deluge/gui/ui/keyboard/layout.h

```diff
@@ -0,0 +1,127 @@
+ * either version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "gui/ui/keyboard/state_data.h"
+#include "model/song/song.h"
+#include "model/instrument/instrument.h"
+#include "model/clip/instrument_clip.h"
+#include "hid/button.h"
+#include <string.h>
+#include <array>
+
+#define INVALID_NOTE -1
+#define MAX_NUM_KEYBOARD_PAD_PRESSES 10
+#define MAX_NUM_ACTIVE_NOTES 10
+
+inline InstrumentClip* currentClip() {
+	return (InstrumentClip*)currentSong->currentClip;
+}
+
+namespace keyboard {
+
+struct PressedPad {
+	uint8_t x;
+	uint8_t y;
+	bool active;
+};
+
+struct NoteState {
+	uint8_t note = 0;
+	uint8_t velocity = 0;
+	int16_t mpeValues[3] = {0};
+	bool generatedNote = false;
+};
+
+constexpr uint8_t kLowestKeyboardNote = 0;
+constexpr uint8_t kHighestKeyboardNote = 12 * 12;
+struct NotesState {
+	uint64_t states[3] = {0};
+	NoteState notes[MAX_NUM_ACTIVE_NOTES] = {0};
+	uint8_t count = 0;
+
+	void enableNote(uint8_t note, uint8_t velocity, bool generatedNote = false, int16_t* mpeValues = nullptr) {
+		if (noteEnabled(note)) {
+			return;
+		}
+		notes[count].note = note;
+		notes[count].velocity = velocity;
+		notes[count].generatedNote = generatedNote;
+		if (mpeValues != nullptr) {
+			memcpy(&notes[count].mpeValues, mpeValues, sizeof(notes[count].mpeValues));
+		}
+
+		states[(note / 64)] |= (1ull << (note % 64));
+		count++;
+	}
+
+	bool noteEnabled(uint8_t note) {
+		uint64_t expectedValue = (1ull << (note % 64));
+		return (states[(note / 64)] & expectedValue) == expectedValue;
+	}
+};
+
+class KeyboardLayout {
+public:
+	KeyboardLayout() {}
+	virtual ~KeyboardLayout() {}
+
+	// Handle inputs
+	virtual void evaluatePads(PressedPad presses[MAX_NUM_KEYBOARD_PAD_PRESSES]) = 0;
+	virtual void handleVerticalEncoder(int offset) = 0; // Shift state not supplied since that function is already taken
+	virtual void handleHorizontalEncoder(int offset, bool shiftEnabled) = 0; // returns weather the scroll had an effect
+	virtual void recalculate() = 0; // This function is called on visibility change and if color offset changes
+
+	// Handle output
+	virtual void renderPads(uint8_t image[][kDisplayWidth + kSideBarWidth][3]) {}
+	virtual void renderSidebarPads(uint8_t image[][kDisplayWidth + kSideBarWidth][3]) {
+		// Empty if not implemented
+		for (int y = 0; y < kDisplayHeight; y++) {
+			memset(image[y][kDisplayWidth], 0, kSideBarWidth * 3);
+		}
+	};
+
+	// Properties
+	virtual char* name() = 0;
+	virtual bool supportsInstrument() { return false; }
+	virtual bool supportsKit() { return false; }
```

Comment:
KeyboardLayouts can support Kit and Instrument type clips at the same time so it would need to be bit encoded. I didn't choose CLIP_TYPE_ since it distinguishes Instrument and Audio type (Audio can't have a Keyboard) and InstrumentType can't encode both at the same time. I think for now it's an okay solution until we have a more cleaned up core and know what makes most sense, do you agree?

Reply:
I meant replacing these separate functions with a single query that's used for this bit of code in `KeyboardScreen::selectLayout`:

```cpp
		if (getActiveInstrument()->type == InstrumentType::KIT && layoutList[nextLayout]->supportsKit()) {
			break;
		}
		else if (getActiveInstrument()->type != InstrumentType::KIT && layoutList[nextLayout]->supportsInstrument()) {
			break;
		}
```

Instead of the separate calls, it could be written as

```
		if (layoutList[nextLayout]->supportsInstrumentType(getActiveInstrument()->type)) {
			break;
		}
```

I think it's OK for all layouts to return `false` for Audio, Midi and CV clips for now -- 2/3 of those make sense to have keyboard support (in fact I'm surprised MIDI clips don't already?) and audio can be special cased if/when someone decides to reuse the keyboard button for audio clips.

Reply:
I'm not sure if I used the correct vocabulary here but I think I read it somewhere else. Instrument refers to Synth, MIDI and CV and the logic is basically NOT Kit. I disregarded Audio completely since it can't open Keyboard view anyway 😅

Reply:
The difference between instrument and clip type is... not super clear to me either. I need to test this PR as it is because today the keyboard layout does work for MIDI and CV tracks but I assume you copied the existing logic so I'm not sure how that works.

Reply:
Yes, Isomorphic and Drum layout work exactly as before except Drums have configurable edge size and isomorphic has a fixed range now (C-2 to C10) which accomodates all underlying clip types. 

If you are okay with it I would like to keep the supportsKit/supportsInstrument logic for now and revisit it if I get a better picture. The overarching logic Kit or NOT Kit was already there and there will be a little more to refactor in keyboard_screen in the long run but we shouldn't need to touch layouts too much from now on.